src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,630 @@
+            return 0;
+        }
+        ret = secp256k1_xonly_ge_serialize(pk_ser, &cache_i.pk);
+        /* Serialization can not fail because the loaded point can not be infinity. */
+        VERIFY_CHECK(ret);
+        pk_ser_ptr = pk_ser;
+    }
+    secp256k1_nonce_function_musig(k, session_id32, seckey, msg32, pk_ser_ptr, extra_input32);
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[0]));
+    VERIFY_CHECK(!secp256k1_scalar_is_zero(&k[1]));
+    secp256k1_musig_secnonce_save(secnonce, k);
+
+    for (i = 0; i < 2; i++) {
+        secp256k1_gej nonce_ptj;
+        secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &nonce_ptj, &k[i]);
+        secp256k1_ge_set_gej(&nonce_pt[i], &nonce_ptj);
+        secp256k1_declassify(ctx, &nonce_pt[i], sizeof(nonce_pt));
+        secp256k1_scalar_clear(&k[i]);
+    }
+    /* nonce_pt can't be infinity because k != 0 */
+    secp256k1_musig_pubnonce_save(pubnonce, nonce_pt);
+    return 1;
+}
+
+static int secp256k1_musig_sum_nonces(const secp256k1_context* ctx, secp256k1_gej *summed_nonces, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    size_t i;
+    int j;
+
+    secp256k1_gej_set_infinity(&summed_nonces[0]);
+    secp256k1_gej_set_infinity(&summed_nonces[1]);
+
+    for (i = 0; i < n_pubnonces; i++) {
+        secp256k1_ge nonce_pt[2];
+        if (!secp256k1_musig_pubnonce_load(ctx, nonce_pt, pubnonces[i])) {
+            return 0;
+        }
+        for (j = 0; j < 2; j++) {
+            secp256k1_gej_add_ge_var(&summed_nonces[j], &summed_nonces[j], &nonce_pt[j], NULL);
+        }
+    }
+    return 1;
+}
+
+int secp256k1_musig_nonce_agg(const secp256k1_context* ctx, secp256k1_musig_aggnonce  *aggnonce, const secp256k1_musig_pubnonce * const* pubnonces, size_t n_pubnonces) {
+    secp256k1_gej aggnonce_ptj[2];
+    secp256k1_ge aggnonce_pt[2];
+    int i;
+
+    ARG_CHECK(aggnonce != NULL);
+    ARG_CHECK(pubnonces != NULL);
+    ARG_CHECK(n_pubnonces > 0);
+
+    if (!secp256k1_musig_sum_nonces(ctx, aggnonce_ptj, pubnonces, n_pubnonces)) {
+        return 0;
+    }
+    for (i = 0; i < 2; i++) {
+        if (secp256k1_gej_is_infinity(&aggnonce_ptj[i])) {
+            /* There must be at least one dishonest signer. If we would return 0
+               here, we will never be able to determine who it is. Therefore, we
+               should continue such that the culprit is revealed when collecting
+               and verifying partial signatures.
+
+               However, dealing with the point at infinity (loading,
+               de-/serializing) would require a lot of extra code complexity.
+               Instead, we set the aggregate nonce to some arbitrary point. This
+               is secure, because it is equivalent to a signer that only inserts
+               themselves to the signing set whenever the the aggregate nonce of
+               all other signers is the point at infinity and then sends some
+               arbitrary point as its nonce.
```

Comment:
Not sure why it's difficult. The pubkey is determined after the nonces are aggregated, so no need for that signer to have pubkey inf. Accepted your suggestion because it is closer to a reduction.